batch/sql/insert_attempt_resources_format_version_lt_3.py

```diff
@@ -0,0 +1,243 @@
+                             '(attempts.batch_id = %s AND attempts.job_id < %s) OR ' \
+                             '(attempts.batch_id = %s AND attempts.job_id = %s AND attempts.attempt_id < %s))'
+                query_args = (resource,
+                              start_batch_id, start_batch_id, start_job_id, start_batch_id, start_job_id, start_attempt_id,
+                              end_batch_id, end_batch_id, end_job_id, end_batch_id, end_job_id, end_attempt_id)
+
+            if size is not None:
+                limit = f'LIMIT {size}'
+            else:
+                limit = ''
+
+            query = f'''
+INSERT INTO attempt_resources (batch_id, job_id, attempt_id, `resource`, quantity)
+SELECT * FROM (
+  SELECT attempts.batch_id, attempts.job_id, attempts.attempt_id, %s, {quantity}
+  FROM attempts FORCE INDEX (PRIMARY)
+  LEFT JOIN jobs ON attempts.batch_id = jobs.batch_id AND attempts.job_id = jobs.job_id
+  LEFT JOIN batches ON attempts.batch_id = batches.id
+  {where_cond}
+  {limit}
+) AS t
+ON DUPLICATE KEY UPDATE quantity = quantity;
+'''
+
+            await tx.just_execute(query, query_args)
+
+        if not quiet:
+            print(f'processed chunk ({start_offset}, {end_offset}) in {time.time() - start_time}s')
+
+        counter.n += 1
+        if counter.n % 500 == 0:
+            print(f'processed {counter.n} chunks')
+
+    return await _process()
+
+
+async def audit_changes(db, expected_n_inserts):
+    result = await db.select_and_fetchone(
+        '''
+SELECT COUNT(*) as count
+FROM attempt_resources
+LEFT JOIN batches ON batches.id = attempt_resources.batch_id
+WHERE batches.`format_version` < 3;
+'''
+    )
+
+    if result['count'] != expected_n_inserts:
+        raise Exception(
+            f'number of attempt resources inserted ({result["count"]})does not match expected value {expected_n_inserts}')
+
+    bad_job_records = db.select_and_fetchall(
+        '''
+SELECT old.batch_id, old.job_id, old.cost, new.cost, ABS(new.cost - old.cost) AS cost_diff
+FROM (
+  SELECT batch_id, job_id, (jobs.msec_mcpu * 0.001 * 0.001) * ((0.01 + ((0.17 * 100 / 30.4375 / 24 + 0.004) / 16) + 0.01) / 3600) AS cost
+  FROM jobs
+  LEFT JOIN batches ON jobs.batch_id = batches.id
+  WHERE batches.`format_version` < 3) AS old
+LEFT JOIN (
+  SELECT attempt_resources.batch_id, attempt_resources.job_id, COALESCE(SUM(GREATEST(COALESCE(end_time - start_time, 0), 0) * quantity * rate), 0) AS cost
+  FROM attempt_resources
+  LEFT JOIN batches ON attempt_resources.batch_id = batches.id
+  LEFT JOIN attempts ON attempt_resources.batch_id = attempts.batch_id AND attempt_resources.job_id = attempts.job_id AND attempt_resources.attempt_id = attempts.attempt_id
+  LEFT JOIN resources ON attempt_resources.resource = resources.resource
+  WHERE batches.`format_version` < 3
+  GROUP BY batch_id, job_id
+) AS new ON new.batch_id = old.batch_id AND new.job_id = old.job_id
+WHERE COALESCE(ABS(new.cost - old.cost), -1) >= 0.001
+LIMIT 100;
```

Comment:
This doesn't really matter but could just be `LIMIT 1`. 100 feels weirdly arbitrary.